substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+					continue
+				};
+				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_output);
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Ignoring ticket over threshold ({:032x} >= {:032x})", ticket_id, ticket_threshold);
+					continue
+				}
+
+				if TicketsData::<T>::contains_key(ticket_id) {
+					debug!(target: LOG_TARGET, "Ignoring duplicate ticket ({:032x})", ticket_id);
+					continue
+				}
+
+				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
+
+				if ticket.signature.ring_vrf_verify(&sign_data, &verifier) {
+					TicketsData::<T>::set(ticket_id, Some(ticket.body));
+					valid_tickets
+						.try_push(ticket_id)
+						.expect("input segment has same length as bounded destination vector; qed");
+				} else {
+					debug!(target: LOG_TARGET, "Proof verification failure");
+				}
+			}
+
+			if !valid_tickets.is_empty() {
+				Self::append_tickets(valid_tickets);
+			}
+
+			Ok(Pays::No.into())
+		}
+
+		/// Plan an epoch config change.
+		///
+		/// The epoch config change is recorded and will be announced at the begin of the
+		/// next epoch together with next epoch authorities information.
+		/// In other words the configuration will be activated one epoch after.
+		/// Multiple calls to this method will replace any existing planned config change that had
+		/// not been enacted yet.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::plan_config_change())]
+		pub fn plan_config_change(
+			origin: OriginFor<T>,
+			config: EpochConfiguration,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(
+				config.redundancy_factor != 0 && config.attempts_number != 0,
+				Error::<T>::InvalidConfiguration
+			);
+			PendingEpochConfigChange::<T>::put(config);
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			let Call::submit_tickets { tickets } = call else {
+				return InvalidTransaction::Call.into()
+			};
+
+			// Discard tickets not coming from the local node or that are not
+			// yet included in a block
+			debug!(
+				target: LOG_TARGET,
+				"Validating unsigned from {} source",
+				match source {
+					TransactionSource::Local => "local",
+					TransactionSource::InBlock => "in-block",
+					TransactionSource::External => "external",
+				}
+			);
+
+			if source == TransactionSource::External {
+				// TODO @davxy: BRAINSTORM this `Local` requirement...
+				// If we only allow these txs on block production, then there is less chance to
+				// submit our tickets if we don't have enough authoring slots.
+				// If we have 0 slots => we have zero chances.
+				// Maybe this is one valid reason to introduce proxies.
+				// In short the question is >>> WHO HAS THE RIGHT TO SUBMIT A TICKET? <<<
+				//  A) The current epoch validators
```

Comment:
This is probably the only thing that makes sense as long as we use unsigned txs no? Otherwise we could introduce DoS vector in the tx pool.